### PR TITLE
Modification about relative position encoding

### PR DESCRIPTION
**Pull request by yaozengwei**
This PR try to simplify the relative position encoding in Conformer. 

In our **existing** Conformer model, we adopt the idea of relative position encoding from Transformer-XL (https://arxiv.org/pdf/1901.02860.pdf).  
The attention weight is composed of content-based score `matrix_ac` and position-based score `matrix_bd`: 
```
# (batch, head, time1, time2)
matrix_ac = (query + bias_u) * key 
# (batch, head, time1, 2*time1-1)
matrix_bd = (query + bias_v) * pos_emb
# (batch, head, time1, time2)
matrix_bd = rel_shift(matrix_bd) 
# (batch, head, time1, time2)
attention_weight = matrix_ac + matrix_bd
```
`matrix_bd` depends on both `query` and `pos_emb`, 
The FLOPs in getting `matrix_bd` is almost **twice** as much as in getting `matrix_ac`.

In the **simplify** method (commit 6fa0ef1e8de430eff07134d16befa11c45e37e4a), position-based score`matrix_bd` only depend on `pos_emb`, and the related FLOPs can be decreased. 
```
# (batch, head, time1, time2)
matrix_ac = query * key
# (1, 2*time1-1, head)
pos_emb = p_weight * pos_emb + p_bias
# (1, head, time1, 2*time1-1), tgt_len is equal to time1
matrix_bd = (pos_emb.transpose(1, 2).unsqueeze(2).repeat(1, 1, tgt_len, 1))
# (1, head, time1, time2)
matrix_bd = rel_shift(matrix_bd)
# (batch, head, time1, time2)
attention_weight = matrix_ac + matrix_bd
```
`p_weight` and `p_bias` are learnable parameters, with shape of `(head, embed_dim)` and `(head, )` , respectively.

The experimental results on librispeech dataset are as follows:
1. Train on train-clean-100 for 30 epochs, decode with greedy search method (epoch-30-avg-10)
* existing method, 6.74 on test-clean, 17.8 on test-other
* simplified method, **6.69** on test-clean, **17.7** on test-other
2. Train on full librispeech for 30 epochs , decode with greedy search method (epoch-30-avg-10)
* existing method, 2.7 on test-clean, 6.25 on test-other
* simplified method, **2.65** on test-clean, **6.18** on test-other

### CODE REVIEW

File: egs/librispeech/ASR/pruned_transducer_stateless7/conformer.py [hunk 1]
```diff
@@ -0,0 +1,1013 @@
...
+            self.pe.size(1) // 2
+            - x.size(1)
+            + 1 : self.pe.size(1) // 2  # noqa E203
+            + x.size(1),
+        ]
+        return self.dropout(x), self.dropout(pos_emb)
+
+
+class RelPositionMultiheadAttention(nn.Module):
+    r"""Multi-Head Attention layer with relative position encoding
```
Comment (danpovey) [inline egs/librispeech/ASR/pruned_transducer_stateless7/conformer.py:408]: We need some clarification about how this was changed.  We can call this  multi-head attention with simplified position encoding.

Comment (yaozengwei) [reply egs/librispeech/ASR/pruned_transducer_stateless7/conformer.py:408]: OK

**Comment by yaozengwei** [comment]
> @danpovey @csukuangfj This is the simplified relative position encoding changes, it needs less computation and produces slightly better results.

**Comment by yaozengwei** [comment]
By using the model profiling tool https://github.com/yaozengwei/model_profiling (like the issue https://github.com/k2-fsa/icefall/issues/385), I compare the relative contribution of different modules in reworked Conformer model to the taining time in two conditions, as shown in the following screenshot: 
(1) with existing relative position encoding (the top split); 
(2) with simplified relative position encoding (the bottom split). 
Screen Shot 2022-07-26 at 22 01 22

**Pull request closed without merging**